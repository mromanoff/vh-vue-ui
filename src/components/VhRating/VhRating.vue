<template>
  <div
    class="VhRating"
    :class="computedClasses"
  >
    <div
      v-for="i in 5"
      :key="i"
      class="VhRating-iconContainer"
      @click="handleIconClick(i)"
      @mouseenter="handleIconHoverEnter(i)"
      @mouseleave="handleIconHoverLeave"
    >
      <vh-icon
        :key="i"
        v-bind="individualIconProps(i)"
        :size="size"
      />
    </div>
  </div>
</template>

<script>
export default {
  name: 'VhRating',
  props: {
    /**
     * Rating component supports the same colors any icon supports
     */
    color: {
      type: String,
      default: 'orange',
    },
    /**
     * Rating provides visual feedback when hovering over icons
     */
    hover: {
      type: Boolean,
      default: false,
    },
    /**
     * Rating component will have half increments
     */
    halfIncrements: {
      type: Boolean,
      default: false,
    },
    /**
     * Rating component will display either stars or cirles
     */
    icon: {
      type: String,
      default: 'star',
      validator: value => {
        return value.match(/(star|circle)/);
      },
    },
    /**
     * Rating will be read only with no events
     */
    readonly: {
      type: Boolean,
      default: false,
    },
    /**
     * Rating takes either xSmall, small, medium, or large size
     */
    size: {
      type: String,
      default: 'medium',
      validator: value => {
        return value.match(/(xSmall|small|medium|large)/);
      },
    },
    /**
     * Rating takes a value from 1 to 5
     */
    value: {
      type: [Number, String],
      default: 0,
      validator: value => {
        return value <= 5;
      },
    },
  },
  data() {
    return {
      hoverIndex: -1,
      internalValue: Number(this.value),
    };
  },
  computed: {
    computedClasses() {
      return {
        'VhRating--readonly': this.readonly,
      };
    },
    isHovering() {
      return this.hover && this.hoverIndex >= 0;
    },
  },
  watch: {
    internalValue(val) {
      if (val !== Number(this.value)) {
        /**
         * input event passes a number representing the rating the user selected
         *
         * @event input
         * @type {Number}
         */
        this.$emit('input', val);
      }
    },
    value(val) {
      this.internalValue = val;
    },
  },
  methods: {
    individualIconProps(i) {
      const PROPS_BANK = {
        circle: {
          empty: { name: 'circle', color: 'grey' },
          half: { name: 'circle', color: this.color },
          full: { name: 'circle', color: this.color },
        },
        star: {
          empty: { name: 'starOutline', color: this.color },
          half: { name: 'starHalf', color: this.color },
          full: { name: 'star', color: this.color },
        },
      };

      if (this.isHovering) {
        const full = Math.floor(this.hoverIndex) >= i;
        return full ? PROPS_BANK[this.icon].full : PROPS_BANK[this.icon].empty;
      }

      if (i <= Math.floor(this.internalValue)) {
        return PROPS_BANK[this.icon].full;
      }

      if (i >= Math.floor(this.internalValue)) {
        if (this.internalValue % 1 > 0.5 && Math.ceil(this.internalValue) === i) {
          return this.halfIncrements ? PROPS_BANK[this.icon].half : PROPS_BANK[this.icon].empty;
        } else {
          return PROPS_BANK[this.icon].empty;
        }
      }
    },
    handleIconClick(i) {
      if (!this.readonly) this.internalValue = i;
    },
    handleIconHoverEnter(i) {
      this.hoverIndex = i;
    },
    handleIconHoverLeave(i) {
      this.hoverIndex = -1;
    },
  },
};
</script>

<style scoped>
/* BLOCK */
.VhRating {
  display: flex;
}

/* ELEMENTS */

.VhRating-iconContainer {
  cursor: pointer;
}

/* MODIFIERS */

.VhRating--readonly {
  pointer-events: none;
}
</style>
