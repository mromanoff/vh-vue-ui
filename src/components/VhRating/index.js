import VhRating from './VhRating.vue';

export { VhRating };
export default VhRating;
