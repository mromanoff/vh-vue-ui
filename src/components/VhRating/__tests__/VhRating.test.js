import { shallowMount } from '@vue/test-utils';
import VhRating from '../VhRating.vue';

describe('VhRating.vue', () => {
  let wrapper;

  it('renders default VhRating', () => {
    wrapper = shallowMount(VhRating, {
      stubs: ['vh-icon'],
    });
    expect(wrapper.props().color).toBe('orange');
    expect(wrapper.props().hover).toBe(false);
    expect(wrapper.props().halfIncrements).toBe(false);
    expect(wrapper.props().icon).toBe('star');
    expect(wrapper.props().readonly).toBe(false);
    expect(wrapper.props().size).toBe('medium');
    expect(wrapper.props().value).toBe(0);
    expect(wrapper).toMatchSnapshot();
  });

  it('renders modified VhRating', () => {
    wrapper = shallowMount(VhRating, {
      propsData: {
        color: 'red',
        hover: true,
        halfIncrements: true,
        icon: 'circle',
        readonly: true,
        size: 'large',
        value: 5,
      },
      stubs: ['vh-icon'],
    });
    expect(wrapper.props().color).toBe('red');
    expect(wrapper.props().hover).toBe(true);
    expect(wrapper.props().halfIncrements).toBe(true);
    expect(wrapper.props().icon).toBe('circle');
    expect(wrapper.props().readonly).toBe(true);
    expect(wrapper.props().size).toBe('large');
    expect(wrapper.props().value).toBe(5);
    expect(wrapper).toMatchSnapshot();
  });
});
