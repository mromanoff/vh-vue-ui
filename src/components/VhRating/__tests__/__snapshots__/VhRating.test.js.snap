// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VhRating.vue renders default VhRating 1`] = `
<div class="VhRating">
  <div class="VhRating-iconContainer">
    <vh-icon-stub size="medium" name="starOutline" color="orange"></vh-icon-stub>
  </div>
  <div class="VhRating-iconContainer">
    <vh-icon-stub size="medium" name="starOutline" color="orange"></vh-icon-stub>
  </div>
  <div class="VhRating-iconContainer">
    <vh-icon-stub size="medium" name="starOutline" color="orange"></vh-icon-stub>
  </div>
  <div class="VhRating-iconContainer">
    <vh-icon-stub size="medium" name="starOutline" color="orange"></vh-icon-stub>
  </div>
  <div class="VhRating-iconContainer">
    <vh-icon-stub size="medium" name="starOutline" color="orange"></vh-icon-stub>
  </div>
</div>
`;

exports[`VhRating.vue renders modified VhRating 1`] = `
<div class="VhRating VhRating--readonly">
  <div class="VhRating-iconContainer">
    <vh-icon-stub size="large" name="circle" color="red"></vh-icon-stub>
  </div>
  <div class="VhRating-iconContainer">
    <vh-icon-stub size="large" name="circle" color="red"></vh-icon-stub>
  </div>
  <div class="VhRating-iconContainer">
    <vh-icon-stub size="large" name="circle" color="red"></vh-icon-stub>
  </div>
  <div class="VhRating-iconContainer">
    <vh-icon-stub size="large" name="circle" color="red"></vh-icon-stub>
  </div>
  <div class="VhRating-iconContainer">
    <vh-icon-stub size="large" name="circle" color="red"></vh-icon-stub>
  </div>
</div>
`;
